README
NAME
    Encode::HanExtra - Extra sets of Chinese encodings

VERSION
    This document describes version 0.20 of Encode::HanExtra, released
    August 21, 2007.

SYNOPSIS
        use Encode;

        # Traditional Chinese
        $euc_tw = encode("euc-tw", $utf8);   # loads Encode::HanExtra implicitly
        $utf8   = decode("euc-tw", $euc_tw); # ditto

        # Simplified Chinese
        $gb18030 = encode("gb18030", $utf8);    # loads Encode::HanExtra implicitly
        $utf8    = decode("gb18030", $gb18030); # ditto

DESCRIPTION
    Perl 5.7.3 and later ships with an adequate set of Chinese encodings,
    including the commonly used "CP950", "CP936" (also known as "GBK"),
    "Big5" (alias for "Big5-Eten"), "Big5-HKSCS", "EUC-CN", "HZ", and
    "ISO-IR-165".

    However, the numbers of Chinese encodings are staggering, and a complete
    coverage will easily increase the size of perl distribution by several
    megabytes; hence, this CPAN module tries to provide the rest of them.

    If you are using Perl 5.8 or later, Encode::CN and Encode::TW will
    automatically load the extra encodings for you, so there's no need to
    explicitly write "use Encode::HanExtra" if you are using one of them
    already.

ENCODINGS
    This version includes the following encoding tables:

      Canonical   Alias                             Description
      -----------------------------------------------------------------------------
      big5-1984   /\b(tca-)?big5-?(19)?84$/i        TCA's original Big5-1984
      big5ext     /\b(cmex-)?big5-?e(xt)?$/i        CMEX's Big5e Extension
      big5plus    /\b(cmex-)?big5-?p(lus)?$/i       CMEX's Big5+ Extension
                  /\b(cmex-)?big5\+$/i
      cccii       /\b(ccag-)?cccii$/i               Chinese Character Code for
                                                    Information Interchange
      cns11643-1  /\bCNS[-_ ]?11643[-_]1$/i         Taiwan's CNS map, plane 1
      cns11643-2  /\bCNS[-_ ]?11643[-_]2$/i         Taiwan's CNS map, plane 2
      cns11643-3  /\bCNS[-_ ]?11643[-_]3$/i         Taiwan's CNS map, plane 3
      cns11643-4  /\bCNS[-_ ]?11643[-_]4$/i         Taiwan's CNS map, plane 4
      cns11643-5  /\bCNS[-_ ]?11643[-_]5$/i         Taiwan's CNS map, plane 5
      cns11643-6  /\bCNS[-_ ]?11643[-_]6$/i         Taiwan's CNS map, plane 6
      cns11643-7  /\bCNS[-_ ]?11643[-_]7$/i         Taiwan's CNS map, plane 7
      cns11643-f  /\bCNS[-_ ]?11643[-_]f$/i         Taiwan's CNS map, plane F
      euc-tw      /\beuc.*tw$/i                     EUC (Extended Unix Character)
                  /\btw.*euc$/i
      gb18030     /\bGB[-_ ]?18030$/i               GBK with Traditional Characters
      unisys      /\bunisys$/i                      Unisys Traditional Chinese
      unisys-sosi1                                  Unisys SOSI1 transport encoding
      unisys-sosi2                                  Unisys SOSI2 transport encoding

    Detailed descriptions are as follows:

    BIG5-1984
        This is the original Big5 encoding made by TCA Taiwan.

    BIG5PLUS
        This encoding, while not heavily used, is an attempt to bring all
        Taiwan's conflicting internal-use encodings together, and fit it as
        an extension to the widely-deployed Big5 range, by CMEX Taiwan.

    BIG5EXT
        The CMEX's second (and less ambitious) try at unifying the most
        commonly used characters not covered by Big5, while not polluting
        out of the 94x94 arragement like BIG5PLUS did.

    CCCII
        The earliest (and most sophisticated) Traditional Chinese encoding,
        with a three-byte raw character map, made in 1980 by the Chinese
        Character Analysis Group (CCAG), used mostly in library systems.

    EUC-TW
        The EUC transport version of "CNS11643" (planes 1-7), the
        comprehensive character set used by the Taiwan government.

    CNS11643-*
        The raw character map extracted from the Unihan database, including
        the plane F which wasn't included in "EUC-TW".

    GB18030
        An extension to GBK, this encoding lists most Han characters (both
        simplified and traditional), as well as some other encodings used by
        other peoples in China.

    UNISYS
        Unisys System's internal Chinese mapping.

NOTES
    If you are looking for ways to transliterate between Simplified and
    Traditional Chinese, please take a look at Encode::HanConvert. Note that
    the direct mapping via Unicode is lossy, and usually doesn't work at
    all.

    Please send me suggestions if you want to see more encoding added, such
    as "BIG5-GCCS" (superseded by "BIG5-HKSCS"). Other suggestions are
    welcome, too.

SEE ALSO
    Encode, Encode::HanConvert

ACKNOWLEDGEMENTS
    Some of the maps here are generated from GNU libiconv's test files, with
    kind permission from Bruno Haible.

    Map for "BIG5PLUS" is generated from the BIG52UCS.TXT file, courtesy of
    CMEX Taiwan (Chinese Microcomputer Extended Foundation,
    <http://www.cmex.org.tw/>).

    Map for "BIG5-1984" is supplied by imacat.

    Map for "CCCII" is supplied by the Koha Taiwan project.

AUTHORS
    Audrey Tang <audreyt@audreyt.org>

COPYRIGHT
    Copyright 2002-2007 by Audrey Tang <audreyt@audreyt.org>.

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 0.22 - 2007-11-08]

* Add Big5-2003, the newest incarnation of Big5 and a Taiwanese national
  standard (as an appendix to CNS11643).

[Changes for 0.21 - 2007-08-22]

* Oops, the CCCII map was broken.  So sowwy. :-/
  Reported by: CPAN Testers.

[Changes for 0.20 - 2007-08-21]

* LICENSING CHANGE: This compilation and all individual files in it
  are now under the permissive "MIT" license.  See the COPYRIGHT
  section in README for the new terms.

* CCCII: Revised, using table supplied by the Koha Taiwan project.

* GB18030: Removed invalid Unicode codepoints.

* Build system upgraded from ExtUtils::MakeMaker to Module::Install.
 
[Changes for 0.10 - 2004-11-29]

* Probe "enc2xs.bat" by default, in addition to "enc2xs".

* Work around heap size problems with Visual C++.

[Changes for 0.09 - 2004-11-19]

* Added the "big5-1984" encoding, supplied by imacat.

* Imported a cleaned-up "cccii" encoding table from http://www.cccii.org.tw/.

* t/1-basic.t now tests encoding as well as decoding.

* Put under SVK version control; reorganized directories into lib/ and ucm/.

* All ucm files now passes ucmlint; large numbers of '|3' were added.

[Changes for 0.08 - 2003-05-18]

* Previous versions were using too much memory; applied
  a patch from Dan Kogai to alleviate it.

* Correct CMEX's URL and name in the documentations.

* Added two Unisys SOSI encodings, courtesy of S-Team inc.

[Changes for 0.06 - 2002-06-03]

* Added CNS11643-* from UniHan.

* Added proper aliases to satisfy Li18nux conformance.

[Changes for 0.05 - 2002-04-21]

* Made the ucm files round-trip safe.

* Added the "big5ext" encoding.

[Changes for 0.04 - 2002-04-18]

* Updated tests.

* Added the "CCCII" encoding from Unihan.

* Changed alias scheme to agree with Encode 1.41+.

* Got rid of compile/encode.h and use enc2xs instead.

[Changes for 0.03 - 2002-03-20]

* Include 2-alias.t to manifest.

[Changes for 0.02 - 2002-03-20]

* First CPAN release.



Makefile.PL
#
# This file is auto-generated by:
# enc2xs version 2.05
# Tue Aug 21 19:01:39 2007
#
use 5.007002;
use strict;
use inc::Module::Install;
use Config;

# Please edit the following to the taste!
my $name   = 'HanExtra';
my %tables = (
    HanExtra_t => [
        map "ucm/$_.ucm",
        (
            qw(big5-1984 big5-2003 big5ext big5plus cccii euc-tw gb18030 unisys),
            map "cns11643-$_",
            (1 .. 7, 'f')
        )
    ],
);

#### DO NOT EDIT BEYOND THIS POINT!
require File::Spec;
my ($enc2xs, $encode_h) = ();
PATHLOOP:
for my $d (@Config{qw/bin sitebin vendorbin/},
    (split /$Config{path_sep}/o, $ENV{PATH}))
{
    for my $f (qw/enc2xs enc2xs5.7.3/) {
        my $path = File::Spec->catfile($d, $f);
        -r $path and $enc2xs = $path and last PATHLOOP;
    }
}
$enc2xs or die "enc2xs not found!";
print "enc2xs is $enc2xs\n";
my %encode_h = ();
for my $d (@INC) {
    my $dir  = File::Spec->catfile($d,   "Encode");
    my $file = File::Spec->catfile($dir, "encode.h");
    -f $file and $encode_h{$dir} = -M $file;
}
%encode_h or die "encode.h not found!";

# find the latest one
($encode_h) = sort { $encode_h{$b} <=> $encode_h{$a} } keys %encode_h;
print "encode.h is at $encode_h\n";

name        'Encode-HanExtra';
all_from    "lib/Encode/$name.pm";
requires    Encode  => 1.41;
auto_provides;
requires_external_cc;
makemaker_args(
    INC          => "-I$encode_h",
    OBJECT       => '$(O_FILES)',
    # OS 390 winges about line numbers > 64K ???
    XSOPT => '-nolinenumbers',
);

sign; WriteAll;

package MY;

sub post_initialize {
    my ($self) = @_;
    my %o;
    my $x = $self->{'OBJ_EXT'};

    # Add the table O_FILES
    foreach my $e (keys %tables) {
        $o{ $e . $x } = 1;
    }
    $o{"$name$x"} = 1;
    $self->{'O_FILES'} = [ sort keys %o ];
    my @files = ("$name.xs");
    $self->{'C'} = ["$name.c"];

    # The next two lines to make MacPerl Happy -- dankogai via pudge
    $self->{SOURCE} .= " $name.c"
      if $^O eq 'MacOS' && $self->{SOURCE} !~ /\b$name\.c\b/;

    # $self->{'H'} = [$self->catfile($self->updir,'encode.h')];
    my %xs;
    foreach my $table (keys %tables) {
        push(@{ $self->{'C'} }, "$table.c");

        # Do NOT add $table.h etc. to H_FILES unless we own up as to how they
        # get built.
        foreach my $ext (qw($(OBJ_EXT) .c .h .exh .fnm)) {
            push(@files, $table . $ext);
        }
    }
    $self->{'XS'} = { "$name.xs" => "$name.c" };
    $self->{'clean'}{'FILES'} .= join(' ', @files);
    open(XS, ">$name.xs") || die "Cannot open $name.xs:$!";
    print XS <<'END';
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#define U8 U8
#include "encode.h"
END
    foreach my $table (keys %tables) {
        print XS qq[#include "${table}.h"\n];
    }
    print XS <<"END";

static void
Encode_XSEncoding(pTHX_ encode_t *enc)
{
 dSP;
 HV *stash = gv_stashpv("Encode::XS", TRUE);
 SV *sv    = sv_bless(newRV_noinc(newSViv(PTR2IV(enc))),stash);
 int i = 0;
 PUSHMARK(sp);
 XPUSHs(sv);
 while (enc->name[i])
  {
   const char *name = enc->name[i++];
   XPUSHs(sv_2mortal(newSVpvn(name,strlen(name))));
  }
 PUTBACK;
 call_pv("Encode::define_encoding",G_DISCARD);
 SvREFCNT_dec(sv);
}

MODULE = Encode::$name  PACKAGE = Encode::$name
PROTOTYPES: DISABLE
BOOT:
{
END
    foreach my $table (keys %tables) {
        print XS qq[#include "${table}.exh"\n];
    }
    print XS "}\n";
    close(XS);
    return "# Built $name.xs\n\n";
}

sub postamble {
    my $self = shift;
    my $dir  = ".";     # $self->catdir('Encode');
    my $str =
"# $name\$(OBJ_EXT) depends on .h and .exh files not .c files - but all written by enc2xs\n";
    $str .= "$name.c : $name.xs ";
    foreach my $table (keys %tables) {
        $str .= " $table.c";
    }
    $str .= "\n\n";
    $str .= "$name\$(OBJ_EXT) : $name.c\n\n";

    foreach my $table (keys %tables) {
        my $numlines    = 1;
        my $lengthsofar = length($str);
        my $continuator = '';
        $str .= "$table.c : Makefile.PL";
        foreach my $file (@{ $tables{$table} }) {
            $str .= $continuator . ' ' . $self->catfile($dir, $file);
            if (length($str) - $lengthsofar > 128 * $numlines) {
                $continuator .= " \\\n\t";
                $numlines++;
            }
            else {
                $continuator = '';
            }
        }
        my $plib = $self->{PERL_CORE} ? '"-I$(PERL_LIB)"' : '';
        my $ucopts = '-"Q"';
        $str .=
          qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
        open(FILELIST, ">$table.fnm")
          || die "Could not open $table.fnm: $!";
        foreach my $file (@{ $tables{$table} }) {
            print FILELIST $self->catfile($dir, $file) . "\n";
        }
        close(FILELIST);
    }
    return $str;
}



lib/Encode/HanExtra.pm
package Encode::HanExtra;
use 5.007003;

our $VERSION = '0.22';
use Encode;
use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

Encode::define_alias( qr/\b(?:tca-)?big5-?(?:19)?84?$/i  => '"big5-1984"'  );
Encode::define_alias( qr/\b(?:cmex-)?big5-?(?:20)?03?$/i => '"big5-2003"'  );
Encode::define_alias( qr/\b(?:cmex-)?big5-?e(?:xt)?$/i   => '"big5ext"'    );
Encode::define_alias( qr/\b(?:cmex-)?big5-?p(?:lus)?$/i  => '"big5plus"'   );
Encode::define_alias( qr/\b(?:cmex-)?big5\+$/i           => '"big5plus"'   );
Encode::define_alias( qr/\b(?:ccag-)?cccii$/i            => '"cccii"'      );
Encode::define_alias( qr/\bunisys$/i                     => '"unisys"'     );
Encode::define_alias( qr/\beuc.*tw$/i                    => '"euc-tw"'     );
Encode::define_alias( qr/\btw.*euc$/i                    => '"euc-tw"'     );
Encode::define_alias( qr/\bGB[-_ ]?18030/i               => '"gb18030"'    );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]1/i         => '"cns11643-1"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]2/i         => '"cns11643-2"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]3/i         => '"cns11643-3"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]4/i         => '"cns11643-4"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]5/i         => '"cns11643-5"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]6/i         => '"cns11643-6"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]7/i         => '"cns11643-7"' );
Encode::define_alias( qr/\bCNS[-_ ]?11643[-_]f/i         => '"cns11643-f"' );

1;

__END__

=head1 NAME

Encode::HanExtra - Extra sets of Chinese encodings

=head1 VERSION

This document describes version 0.22 of Encode::HanExtra, released
November 8, 2007.

=head1 SYNOPSIS

    use Encode;

    # Traditional Chinese
    $euc_tw = encode("euc-tw", $utf8);   # loads Encode::HanExtra implicitly
    $utf8   = decode("euc-tw", $euc_tw); # ditto

    # Simplified Chinese
    $gb18030 = encode("gb18030", $utf8);    # loads Encode::HanExtra implicitly
    $utf8    = decode("gb18030", $gb18030); # ditto

=head1 DESCRIPTION

Perl 5.7.3 and later ships with an adequate set of Chinese encodings,
including the commonly used C<CP950>, C<CP936> (also known as C<GBK>),
C<Big5> (alias for C<Big5-Eten>), C<Big5-HKSCS>, C<EUC-CN>, C<HZ>, and
C<ISO-IR-165>.

However, the numbers of Chinese encodings are staggering, and a complete
coverage will easily increase the size of perl distribution by several
megabytes; hence, this CPAN module tries to provide the rest of them.

If you are using Perl 5.8 or later, L<Encode::CN> and L<Encode::TW> will
automatically load the extra encodings for you, so there's no need to
explicitly write C<use Encode::HanExtra> if you are using one of them
already.

=head1 ENCODINGS

This version includes the following encoding tables:

  Canonical   Alias                             Description
  -----------------------------------------------------------------------------
  big5-1984   /\b(tca-)?big5-?(19)?84$/i        TCA's original Big5-1984
  big5-2003   /\b(cmex-)?big5-?(20)?03$/i       Big5-2003 (national standard)
  big5ext     /\b(cmex-)?big5-?e(xt)?$/i        CMEX's Big5e Extension
  big5plus    /\b(cmex-)?big5-?p(lus)?$/i       CMEX's Big5+ Extension
              /\b(cmex-)?big5\+$/i
  cccii       /\b(ccag-)?cccii$/i               Chinese Character Code for
                                                Information Interchange
  cns11643-1  /\bCNS[-_ ]?11643[-_]1$/i         Taiwan's CNS map, plane 1
  cns11643-2  /\bCNS[-_ ]?11643[-_]2$/i         Taiwan's CNS map, plane 2
  cns11643-3  /\bCNS[-_ ]?11643[-_]3$/i         Taiwan's CNS map, plane 3
  cns11643-4  /\bCNS[-_ ]?11643[-_]4$/i         Taiwan's CNS map, plane 4
  cns11643-5  /\bCNS[-_ ]?11643[-_]5$/i         Taiwan's CNS map, plane 5
  cns11643-6  /\bCNS[-_ ]?11643[-_]6$/i         Taiwan's CNS map, plane 6
  cns11643-7  /\bCNS[-_ ]?11643[-_]7$/i         Taiwan's CNS map, plane 7
  cns11643-f  /\bCNS[-_ ]?11643[-_]f$/i         Taiwan's CNS map, plane F
  euc-tw      /\beuc.*tw$/i                     EUC (Extended Unix Character)
              /\btw.*euc$/i
  gb18030     /\bGB[-_ ]?18030$/i               GBK with Traditional Characters
  unisys      /\bunisys$/i                      Unisys Traditional Chinese
  unisys-sosi1                                  Unisys SOSI1 transport encoding
  unisys-sosi2                                  Unisys SOSI2 transport encoding

Detailed descriptions are as follows:

=over 4

=item BIG5-1984

This is the original Big5 encoding made by TCA Taiwan.

=item BIG5-2003

This revised encoding is now national standard, as an appendix of CNS11643.

=item BIG5PLUS

This encoding, while not heavily used, is an attempt to bring all Taiwan's
conflicting internal-use encodings together, and fit it as an extension to
the widely-deployed Big5 range, by CMEX Taiwan.

=item BIG5EXT

The CMEX's second (and less ambitious) try at unifying the most commonly
used characters not covered by Big5, while not polluting out of the 94x94
arragement like BIG5PLUS did.

=item CCCII

The earliest (and most sophisticated) Traditional Chinese encoding, with a
three-byte raw character map, made in 1980 by the Chinese Character Analysis
Group (CCAG), used mostly in library systems.

=item EUC-TW

The EUC transport version of C<CNS11643> (planes 1-7), the comprehensive
character set used by the Taiwan government.

=item CNS11643-*

The raw character map extracted from the Unihan database, including the
plane F which wasn't included in C<EUC-TW>.

=item GB18030

An extension to GBK, this encoding lists most Han characters (both simplified
and traditional), as well as some other encodings used by other peoples in
China.

=item UNISYS

Unisys System's internal Chinese mapping.

=back

=head1 NOTES

If you are looking for ways to transliterate between Simplified and
Traditional Chinese, please take a look at L<Encode::HanConvert>. Note
that the direct mapping via Unicode is lossy, and usually doesn't work
at all.

Please send me suggestions if you want to see more encoding added, such
as C<BIG5-GCCS> (superseded by C<BIG5-HKSCS>).  Other suggestions are welcome,
too.

=head1 SEE ALSO

L<Encode>, L<Encode::HanConvert>

=head1 ACKNOWLEDGEMENTS

Some of the maps here are generated from GNU libiconv's test files,
with kind permission from Bruno Haible.

Map for C<BIG5PLUS> is generated from the F<BIG52UCS.TXT> file,
courtesy of CMEX Taiwan (Chinese Microcomputer Extended Foundation,
L<http://www.cmex.org.tw/>).

Map for C<BIG5-1984> is supplied by imacat.

Map for C<CCCII> is supplied by the Koha Taiwan project.

=head1 AUTHORS

Audrey Tang E<lt>audreyt@audreyt.orgE<gt>

=head1 COPYRIGHT

Copyright 2002-2007 by Audrey Tang E<lt>audreyt@audreyt.orgE<gt>.

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
